gap/attr.gi

```diff
@@ -128,18 +135,29 @@ function(digraph)
     if not IsNullDigraph(subdigraph)
         and not IsBipartiteDigraph(subdigraph) then
       if IsCompleteDigraph(DigraphSymmetricClosure(subdigraph)) then
-        if n > chrom then
+        if chrom = fail or n > chrom then
           chrom := n;
         fi;
       else
-        for i in [3 .. n - 1] do
-          if DigraphColoring(subdigraph, i) <> fail then
-            if i > chrom then
-              chrom := i;
+        upper := RankOfTransformation(DigraphColouring(subdigraph), nr);
+        if chrom = fail or upper > chrom then
+          upper := Minimum(upper, nr - 1);
```

Comment:
The problem with `n` vs `nr` is causing this kind of bug:

`gr := DigraphFromGraph6String("KmKk~K??G@_@");`

This is giving a chromatic number of 11, though the correct answer is 4. That's because it's trying to colour one of its components (that has 4 vertices) using 10 colours, which returns `fail` because not all 10 colours can be used for 4 vertices! That's because `upper` was originally set to 11, from the greedy colouring.